File: packages/remote-worker-rs/src/commands/git_call.rs
use std::time::Duration;

use crate::protocol::{self, WsRequest};

pub async fn run(req: &WsRequest) -> String {
	let request_id = &req.request_id;

	let repo_path = match req.payload.get("repo_path").and_then(|v| v.as_str()) {
		Some(p) if !p.is_empty() => p.to_string(),
		_ => return protocol::error(request_id, "repo_path must be a non-empty string"),
	};

	let args: Vec<String> = match req.payload.get("args").and_then(|v| v.as_array()) {
		Some(arr) => {
			let mut result = Vec::new();
			for v in arr {
				match v.as_str() {
					Some(s) => result.push(s.to_string()),
					None => return protocol::error(request_id, "args must be a string[]"),
				}
			}
			result
		}
		None => return protocol::error(request_id, "args must be a string[]"),
	};

	let resolved = if std::path::Path::new(&repo_path).is_absolute() {
		std::path::PathBuf::from(&repo_path)
	} else {
		std::env::current_dir().unwrap_or_default().join(&repo_path)
	};

	if !resolved.exists() {
		return protocol::error(
			request_id,
			&format!("Repository path does not exist: {}", resolved.display()),
		);
	}

	eprintln!("[git] {:?}", args);

	let mut retries = 3u32;
	let mut delay_ms = 100u64;

	loop {
		match run_git(&resolved, &args).await {
			Ok(stdout) => return protocol::success(request_id, stdout.into()),
			Err(msg) if msg.contains(".git/index.lock': File exists") && retries > 0 => {
				tokio::time::sleep(Duration::from_millis(delay_ms)).await;
				retries -= 1;
				delay_ms *= 2;
			}
			Err(msg) => return protocol::error(request_id, &msg),
		}
	}
}

async fn run_git(cwd: &std::path::Path, args: &[String]) -> Result<String, String> {
	let output = tokio::process::Command::new("git")
		.args(args)
		.current_dir(cwd)
		.output()
		.await
		.map_err(|e| e.to_string())?;

	if output.status.success() {
		Ok(String::from_utf8_lossy(&output.stdout).into_owned())
	} else {
		let stdout = String::from_utf8_lossy(&output.stdout);
		let stderr = String::from_utf8_lossy(&output.stderr);
		Err(format!("{stdout}{stderr}"))
	}
}

File: packages/server/src/commands/GitCall.ts
import {resolve, isAbsolute} from 'path';
import {existsSync} from 'fs';
import {getAgentEnv} from './SshAgentInit';
import type {IWsMessage} from '../types';

function validateRepoPath(repoPath: unknown): string {
	if (typeof repoPath !== 'string' || !repoPath) {
		throw new Error('repo_path must be a non-empty string');
	}

	const resolved = isAbsolute(repoPath) ? repoPath : resolve(repoPath);

	if (!existsSync(resolved)) {
		throw new Error(`Repository path does not exist: ${resolved}`);
	}

	return resolved;
}

export async function run(ws: {send: (msg: string) => void}, data: IWsMessage): Promise<void> {
	const {requestId, repo_path, args} = data;

	if (!Array.isArray(args) || !args.every((a): a is string => typeof a === 'string')) {
		ws.send(JSON.stringify({requestId, status: 'error', message: 'args must be a string[]'}));
		return;
	}

	let resolvedPath: string;

	try {
		resolvedPath = validateRepoPath(repo_path);
	}
	catch (e: unknown) {
		const message = e instanceof Error ? e.message : 'Invalid repository path';
		ws.send(JSON.stringify({requestId, status: 'error', message}));
		return;
	}

	console.log('[git]', args);

	const executeGit = async (): Promise<string> => {
		const proc = Bun.spawn(['git', ...args], {
			cwd: resolvedPath,
			env: getAgentEnv(ws),
			stdout: 'pipe',
			// Bun defaults stderr to "inherit" — capture it so git errors
			// (push rejections, conflicts, etc.) reach the client.
			stderr: 'pipe',
		});

		const [stdout, stderr] = await Promise.all([
			new Response(proc.stdout).text(),
			new Response(proc.stderr).text(),
		]);
		const exitCode = await proc.exited;

		if (exitCode === 0) {
			// Hooks write diagnostics to stderr; return it for `commit` so the
			// client can show hook output on success. git's own commit summary
			// goes to stdout and is redundant (the UI refreshes), so it's dropped
			// here to avoid a popup on every plain commit. Other commands keep
			// clean stdout for parsing.
			if (args[0] === 'commit') {
				return stderr;
			}

			return stdout;
		}

		const combined = (stdout + stderr).trim();

		throw new Error(combined || `git ${args.join(' ')} exited with code ${exitCode}`);
	};

	let retries = 3;
	let delay = 100;

	while (true) {
		try {
			const result = await executeGit();
			ws.send(JSON.stringify({requestId, status: 'success', data: result}));
			break;
		}
		catch (e: unknown) {
			const message = e instanceof Error ? e.message : 'Unknown git error';

			if (message.includes(`.git/index.lock': File exists`) && retries) {
				await new Promise(r => setTimeout(r, delay));
				retries--;
				delay *= 2;
				continue;
			}

			ws.send(JSON.stringify({requestId, status: 'error', message}));
			break;
		}
	}
}
